\section{Introduction}
\label{sec:introduction}

\emph{Clustering} is the task of partitioning a set of objects, such that each
partition contains objects that are similar to each other and objects in
separate partitions are dissimilar, according to some well-defined notion of
similarity. These partitions are called \emph{clusters}.

Clustering huge amounts of DNA/RNA sequences (up to \num{500} million strings of
\num{500}--\num{1500} characters each) is computationally hard, and the cost of sequencing,
i.e. determining DNA and RNA data, has dropped so low, and consequently the
amount of data has grown so large, that the available algorithms and hardware
can no longer keep up~\cite{rothberg}.

There are not many available algorithms and tools for efficient clustering of
sequencing data. The one tool which is currently doing the job,
\texttt{USEARCH} \\ 
\cite{edgar,usearch}, is closed-source and although the \num{32}-bit
version is available for free, the usable memory is limited to \num{4} GB and the
\num{64}-bit version is costly.

As the amount of sequencing data grows, clustering becomes increasingly more
relevant as it can be used to reduce a large set of sequences to a much smaller
set of representative sequences, called \emph{centroids}, which each represents
a single cluster.

This project is concerned with the clustering of DNA and RNA sequences. The
clustering method described in this thesis strives to assign each sequence to
a cluster based on the similarity between the sequence and the centroid. The
clustering method does not guarantee a minimal number of clusters, although it
tries to minimize it, as the search for a centroid is not exhaustive, nor does
it guarantee to find the most similar centroid for each sequence.

This project researches the possibilities for creating an open-source tool that
can match the performance of \texttt{USEARCH}. We present a solution and
implementation that uses the concept of \emph{k-mer counting}, i.e. essentially
comparing the distribution of all possible substrings of a fixed length $k$ in
each sequence, as an approximation for measuring sequence similarity.
Furthermore, it utilizes the $k$-mers occurring in the centroids for improving
the search for a centroid that is likely to be a good representative for a
given sequence.

Earlier works have been very focused on techniques based on sequence alignment,
with \texttt{USEARCH} being one such example. However, a clustering algorithm
with a similarity measure based on $k$-mers, can possible give an improved
performance in terms of speed, while still being a sufficiently good
approximation.

Our program, named \texttt{klust}, has been implemented in \texttt{C++}, is
hosted on GitHub and is available at \url{https://github.com/Rathcke/klust}.

The reader of this text is expected to have a basic knowledge of computer
science, corresponding to around two or three years of undergraduate studies.


\subsection{Defining the clustering problem}

The clustering problem that this project aims to solve is a partitioning of
the sequences such that the similarities between a given centroid and all
sequences in the cluster represented by that sequence is greater than or equal
to the specified threshold similarity. There is, however, no guarantee that
the centroid sequences will have a lower mutual similarity than the similarity
threshold although this should preferably be the case. Figure
\ref{fig:clustering_concept} illustrates this concept of \emph{centroid based}
clustering.

\begin{figure}[h!]
  \centering
  \def\svgwidth{0.9\columnwidth}
  \import{graphics/}{Clustering.pdf_tex}
  \caption{Illustration of centroid based clustering. The green sequences
    belong to the cluster represented by the leftmost, black centroid and the
    blue sequences belong to the rightmost cluster. The sequence in the
    intersection between the circles belong to only the left cluster although
    it could possibly belong to the right one instead (as indicated by
    the dotted arrow).}
  \label{fig:clustering_concept}
\end{figure}

Let $C$ be a set of pairs containing a centroid and the set of sequences
belonging to the cluster represented by that centroid. Let $sim$ denote the
similarity measure and let $t$ be a given threshold similarity. Then a certain
partitioning of sequences is a solution to the above clustering problem if it
satisfies the following property: \[   \forall\, (c,S) \in C:\; \forall\, s
\in S: sim(s,c) \geq t \]
